### PR TITLE
Unix: copy buffer before passing off to client

### PR DESCRIPTION
**Pull request by reynir**
Originally reported by @rand00.

### CODE REVIEW

File: src/stack-unix/udpv4v6_socket.ml [hunk 1]
```diff
@@ -191,6 +190,11 @@ let listen t ~port callback =
                            | Some v4 -> Ipaddr.V4 v4
                          in
                          let dst = Ipaddr.(V6 V6.unspecified) in (* TODO *)
+                         let buf =
+                           let b = Cstruct.create_unsafe len in
+                           Cstruct.blit buf 0 b 0 len;
+                           b
+                         in
                          callback ~src ~dst ~src_port buf
```
Comment (hannesm) [inline src/stack-unix/udpv4v6_socket.ml:198]: What I understood from the report #501 is that here this callback in the Unix stack is run synchronously with Lwt_cstruct.recvfrom (so there won't be multiple callbacks executed concurrently) -- while in the Mirage stack due to every mirage-net-* implementation calling "Lwt.async" for the callback of each incoming packet, the MirageOS UDP callback is asynchronous (i.e. there can be multiple callbacks concurrently).

I still don't quite understand what the desired semantics should be -- maybe the callback should not involve "io / Lwt.t" at all? Maybe it should be all sync? My intuition is that the "Lwt.async" for each incoming packet is not optimal from the performance and scheduling perspective.

Comment (reynir) [reply src/stack-unix/udpv4v6_socket.ml:198]: I think this PR is orthogonal to #501. @rand00 also reported a buffer ownership bug on IRC (but no issue AFAICT).

Comment (hannesm) [reply src/stack-unix/udpv4v6_socket.ml:198]: ah, ok :)

File: src/stack-unix/udpv4v6_socket.ml [hunk 2]
```diff
@@ -191,6 +190,11 @@ let listen t ~port callback =
                            | Some v4 -> Ipaddr.V4 v4
                          in
                          let dst = Ipaddr.(V6 V6.unspecified) in (* TODO *)
+                         let buf =
+                           let b = Cstruct.create_unsafe len in
+                           Cstruct.blit buf 0 b 0 len;
+                           b
+                         in
```
Comment (hannesm) [inline src/stack-unix/udpv4v6_socket.ml:197]: maybe use Cstruct.copy?

Comment (reynir) [reply src/stack-unix/udpv4v6_socket.ml:197]: Cstruct.copy is Cstruct.t -> int -> int -> string, that is, `Cstruct.copy t off len` is the same as `Cstruct.to_string t ~off ~len` :/ I didn't see a function to copy a cstruct (even though Cstruct.copy would be an obvious name for such a function)

Comment (reynir) [reply src/stack-unix/udpv4v6_socket.ml:197]: Hi @palainp, indeed, `Cstruct.sub_copy` does what I want, but it only exists in unreleased cstruct. I can put a comment in the code to use `Cstruct.sub_copy` in the future.

Comment (palainp) [reply src/stack-unix/udpv4v6_socket.ml:199]: Oh sorry, didn't checked the last release date :(

Comment (dinosaure) [reply src/stack-unix/udpv4v6_socket.ml:199]: I will cut a release of `cstruct` as soon as I can but some improvements are not merged yet.

**Review by dinosaure — APPROVED** [review]

**Pull request merged**